luarules/gadgets/game_tax_resource_sharing.lua

```diff
@@ -56,27 +69,64 @@ function gadget:AllowResourceTransfer(senderTeamId, receiverTeamId, resourceType
 	local maxShare = rStor * rShare - rCur
 
-	local taxedAmount = math.min((1-sharingTax)*amount, maxShare)
-	local totalAmount = taxedAmount / (1-sharingTax)
-	local transferTax = totalAmount * sharingTax
+	-- Prevent negative maxShare
+	maxShare = math.max(0, maxShare)
+
+	-- Calculate the actual amount to transfer, respecting maxShare
+	local transferAmount = math.min(amount, maxShare)
+
+	local currentSharingTax = sharingTax
+	local actualSentAmount = 0  -- How much sender actually loses
+	local actualReceivedAmount = 0 -- How much receiver actually gains
+	local currentCumulative = 0 -- Declare here to ensure scope
+
+	-- Apply cumulative threshold logic only for metal
+	if resourceName == 'metal' and metalTaxThreshold > 0 then
+		currentCumulative = cumulativeMetalSent[senderTeamId] or 0 -- Assign value here
+
+		local allowanceRemaining = math.max(0, metalTaxThreshold - currentCumulative)
+		local untaxedPortion = math.min(transferAmount, allowanceRemaining)
+		local taxablePortion = transferAmount - untaxedPortion
+
+		if taxablePortion > 0 then
+			-- Apply tax only to the taxable portion
+			local taxedPortionReceived = taxablePortion * (1 - sharingTax)
+			local taxedPortionSent = taxablePortion / (1 - sharingTax) -- Assuming tax isn't 100%
+			if sharingTax == 1 then taxedPortionSent = taxablePortion end -- Handle 100% tax case
+
+			actualReceivedAmount = untaxedPortion + taxedPortionReceived
+			actualSentAmount = untaxedPortion + taxedPortionSent
+		else
+			-- Entire transfer is within the remaining allowance, no tax
+			actualReceivedAmount = untaxedPortion -- which equals transferAmount
+			actualSentAmount = untaxedPortion     -- which equals transferAmount
+			currentSharingTax = 0 -- Log purpose
+		end
+
+	else -- Energy transfer OR Metal Threshold is 0
+		actualReceivedAmount = transferAmount * (1 - currentSharingTax)
+		actualSentAmount = actualReceivedAmount / (1 - currentSharingTax) -- Assuming tax isn't 100%
+		if currentSharingTax == 1 then actualSentAmount = transferAmount end -- Handle 100% tax case
+	end
 
-	Spring.SetTeamResource(receiverTeamId, resourceName, rCur+taxedAmount)
-	local sCur, _, _, _, _, _ = Spring.GetTeamResources(senderTeamId, resourceName)
-	Spring.SetTeamResource(senderTeamId, resourceName, sCur-totalAmount)
+	-- Ensure we don't send more than originally intended due to tax calculation edge cases / maxShare limit
+	actualSentAmount = math.min(actualSentAmount, amount) 
+	actualReceivedAmount = math.min(actualReceivedAmount, transferAmount)
 
-	-- Block the original transfer
-	return false
-end
+	-- Perform the transfer
+	Spring.SetTeamResource(receiverTeamId, resourceName, rCur + actualReceivedAmount)
+	local sCur, _, _, _, _, _ = Spring.GetTeamResources(senderTeamId, resourceName)
+	Spring.SetTeamResource(senderTeamId, resourceName, sCur - actualSentAmount)
 
-function gadget:AllowUnitTransfer(unitID, unitDefID, oldTeam, newTeam, capture)
```

Comment:
did you intent to delete this?

Reply:
yes it was redundant with another widget doing the exact same logic in the same hook.